ENH: .equals for Extension Arrays

dwhu opened the pull request:
- [x] closes #27081
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

--- Jan 3 10:00 PST

Hey team,

Looking for a bit of feedback on this API design and appropriate testing for this PR before I get too deep into it. Could you provide some guidance regarding how to appropriately handle updating this test? https://github.com/pandas-dev/pandas/blob/6f03e76f9d47ecfcfdd44641de6df1fc7dd57a01/pandas/tests/extension/base/ops.py#L162 It appears to contradict the issue #27081.

Appreciate the help,
Dave

dwhu committed:
ENH: .equals for Extension Arrays

charlesdong1991 commented:
thanks for your contribution! @dwhu 

in general here, if it is `NotImplemted`, and you implement some new functionalities, you could remove this one, and replace it with the corresponding tests to test your new implementations!

Good luck! and hope this helps

dwhu commented:
> in general here, if it is `NotImplemted`, and you implement some new functionalities, you could remove this one, and replace it with the corresponding tests to test your new implementations!

@charlesdong1991 - Sounds good. Wanted to make sure I wasn't missing something or breaking a behavior contract by adding this behavior.

dwhu committed:
ENH: Updating eq and ne methods for extension arrays.

Adding __eq__ to ExtensionArray Abstract method doc string.

Adding ne implementation to EA base class.

Also removing other implementations.

Updating EA equals method and adding tests.

GH-27081

TomAugspurger commented on pandas/core/arrays/base.py:
```diff
@@ -350,6 +351,38 @@ def __iter__(self):
         for i in range(len(self)):
             yield self[i]
 
+    def __eq__(self, other: ABCExtensionArray) -> bool:
```
These will need to be added to the list of abstract methods in the "Notes" section of the class docstring.

dwhu replied:
Done

TomAugspurger commented on pandas/core/arrays/base.py:
```diff
@@ -350,6 +351,38 @@ def __iter__(self):
+        bool
+        """
+
+        raise AbstractMethodError(self)
```
Can this be `return ~(self == other)`?

dwhu replied:
Done.

TomAugspurger commented on pandas/core/arrays/base.py:
```diff
@@ -657,6 +690,23 @@ def searchsorted(self, value, side="left", sorter=None):
+            Whether the arrays are equivalent.
+
+        """
+        return ((self == other) | (self.isna() == other.isna())).all()
```
I think this also needs to ensure that the type of `other` matches the type of `self`. We don't want `pd.array([1, 2]).equals(np.array([1, 2]))` to be True (I think).

dwhu replied:
Done.

TomAugspurger commented on pandas/core/arrays/boolean.py:
```diff
@@ -314,6 +314,20 @@ def __getitem__(self, item):
+        if not isinstance(other, BooleanArray):
+            return NotImplemented
+        return (
+            hasattr(other, "_data")
```
BooleanArray should already have a `__eq__` method (see `create_comparison_method` Was this needed?

dwhu replied:
Not needed. Missed that in my WIP PR.

TomAugspurger reviewed (commented):
I think most of the `__eq__` implementations you've added can be removed.

@jschendel is in the middle of a PR fixing `IntervalArray.__eq__`, so ignore that for now if it's causing failures.

We'll need some base tests in `pandas/tests/extensions/base/ops.py` ensuring that `__eq__` and `__ne__` are implemented correctly. We'll want a few tests in `tests/extensions/base/methods.py` ensuring that `.equals` is tested. These tests are a bit tricky to write, so let us know if you have issues.

dwhu committed:
Removing interval.py's __eq__ implementation due to conflict with @jschendel change.

TomAugspurger commented on pandas/core/arrays/base.py:
```diff
@@ -657,6 +691,25 @@ def searchsorted(self, value, side="left", sorter=None):
+            Whether the arrays are equivalent.
+
+        """
+        return isinstance(other, self.__class__) and (
```
Perhaps `type(self) == type(other)`? It's not clear to me whether a subclass should compare `.equals` to a parent class. Thoughts @jreback?

TomAugspurger replied:
Will want some length checking here too.

dwhu replied:
Added `type(self) == type(other)` and length checks.

TomAugspurger commented on pandas/core/arrays/integer.py:
```diff
@@ -376,6 +376,15 @@ def __getitem__(self, item):
+    def __eq__(self, other):
+        return (
+            isinstance(other, IntegerArray)
+            and hasattr(other, "_data")
```
We already have `IntegerArray.__eq__` on master. This shouldn't be necessary.

TomAugspurger commented on pandas/tests/extension/base/methods.py:
```diff
@@ -358,3 +358,18 @@ def test_repeat_raises(self, data, repeats, kwargs, error, msg, use_numpy):
+        ser = pd.Series(cls._from_sequence(data, dtype=data.dtype))
+        na_ser = pd.Series(cls._from_sequence([na_value], dtype=data.dtype))
+
+        assert data.equals(data)
```
make all these `assert <condition> is True`, to ensure that `True` or `False` is actually returned

dwhu replied:
I ran into a funny issue doing this: https://github.com/numpy/numpy/issues/5942
I'm going to keep this as is now to keep the test consistent within itself.

`.equals()` returns a `np.bool_` type. As result when you compare it to bool type `True` using `is` (i.e. `some_np_bool_True_val is True == False` 

Coercing `np.bool_` to `bool` using the `bool()` constructor did not work either. `== True` causes PEP8 errors asking me to use `assert <condition>` or `assert <condition> is True`.

Open to suggestions if you have any.

TomAugspurger commented on pandas/tests/extension/base/methods.py:
```diff
@@ -358,3 +358,18 @@ def test_repeat_raises(self, data, repeats, kwargs, error, msg, use_numpy):
+        assert not na_ser.equals(ser)
+        assert not ser.equals(na_ser)
+        assert not ser.equals(0)
+        assert not na_ser.equals(0)
```
Add test for other inputs like an ndarray.

Add a test for unequal length things.

dwhu replied:
Done.

TomAugspurger commented on pandas/tests/extension/base/ops.py:
```diff
@@ -132,10 +132,8 @@ class BaseComparisonOpsTests(BaseOpsUtil):
     def _compare_other(self, s, data, op_name, other):
         op = self.get_op_from_name(op_name)
         if op_name == "__eq__":
-            assert getattr(data, op_name)(other) is NotImplemented
```
We definitely don't want to remove this :) You'll need to revert it and whatever change required it :)

jorisvandenbossche replied:
@TomAugspurger I think this is actually correct to change this. As far as I understand, we only tested this as the base version (if you don't define an `__eq__` on your class, python will return this, I assume). But now we changed the base class to actually have a `__eq__` but that raise an AbstractError, so it needs to always be implemented, and thus can return something here that is not NotImplemented.

TomAugspurger replied:
What's `other` here? If it's a Series / DataFrame we need to make sure to return `NotImplemented` right? I thought that was the intent of this test, though I might misunderstand.

jorisvandenbossche replied:
`data` is the EA, and other can eg be a scalar. 

There is a separate test just below (that I added back) that explicitly tests that ``NotImplemented`` is returned for `EA.__eq__(Series/Index)`

TomAugspurger commented on pandas/tests/extension/base/ops.py:
```diff
@@ -158,13 +156,3 @@ def test_compare_array(self, data, all_compare_operators):
         other = pd.Series([data[0]] * len(data))
         self._compare_other(s, data, op_name, other)
-
-    def test_direct_arith_with_series_returns_not_implemented(self, data):
```
Undo this change.

jorisvandenbossche replied:
> Same as the other comment. The whole PR is intended to break this test since we implement __eq__ and __ne__.

The reason this needs to be kept is that, although `__eq__` is a required method on an ExtensionArray now, we still require it to return NotImplemented when `other` is a Series or Index.

dwhu committed:
ENH: Making EA eq and ne typed as Any.

GH-27081

dwhu committed:
ENH: Adding default implementation to ExtensionArray equals() and tests.

GH-27081

dwhu committed:
Merge remote-tracking branch 'upstream/master' into gh-27081

WillAyd commented:
@dwhu looks like CI is red - can you take a look and fix up the merge conflict?

WillAyd commented:
Closing as I think this has gone stale but ping if you'd like to pick back up. Thanks!

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into gh-27081

jorisvandenbossche commented:
@dwhu I am going to look into further updating this, if that's OK with you

jorisvandenbossche committed:
correct __eq/ne__ to be element-wise

jorisvandenbossche committed:
fix equals implementation (& instead of ==)

jorisvandenbossche committed:
base tests

jorisvandenbossche committed:
ensure to dispatch Series.equals to EA.equals

jorisvandenbossche commented:
OK, I updated this with the following things:

- Made `__eq__` into an abstract method so it is required to implement (another option could be to `return NotImplemented` as default, but that's actually wrong, as it results in not doing element-wise comparison for `EA == EA`, but just returning False)
- Fixed the `equals` imlementation: it became a bit longer to ensure we check for BooleanArray (if the `==` comparison returns a nullable boolean array, we need to fill NAs with False)
- Added `ExtensionBlock.equals` (overriding the base Block method) to ensure we dispatch to the new EA method
- A bit of clean-up of the tests

Should be ready for a new review now @TomAugspurger @jbrockmendel

TomAugspurger commented on pandas/core/arrays/base.py:
```diff
@@ -333,6 +335,18 @@ def __iter__(self):
 
+    def __eq__(self, other: Any) -> Union[np.ndarray, "ExtensionArray"]:
+        """
+        Return for `self == other` (element-wise equality).
```
I think we should have some guidance here that if `other` is a Series then you should return NotImplemented.

jbrockmendel replied:
could use ops.common.unpack_zerodim_and_defer

jorisvandenbossche replied:
> could use ops.common.unpack_zerodim_and_defer

This method is to be implemented by EA authors, so those can't use that helper (unless we expose somewhere a public version of this). 

(we could of course use that for our own EAs, but this PR is not changing any existing `__eq__` implementation at the moment)

jorisvandenbossche replied:
> I think we should have some guidance here that if other is a Series then you should return NotImplemented.

Added a comment about that

TomAugspurger commented on pandas/core/arrays/base.py:
```diff
@@ -682,6 +696,33 @@ def searchsorted(self, value, side="left", sorter=None):
+            Whether the arrays are equivalent.
+
+        """
+        if not type(self) == type(other):
+            return False
```
We should verify that this is the behavior we want. Namely

1. other array-likes are not equivalent, even if they are all equal.
2. *subclasses* are not equivalent, even if they are all equal.

The first seems fine. Not sure about the second.

jorisvandenbossche replied:
I was planning to open an issue about this after this PR (and keep it strict here), because this is right now a bit inconsistent within pandas, and might require a more general discussion / clean-up (eg Series.equals is more strict (requires same dtype) than Index.equals ...) 
But we can certainly also have the discussion here.

TomAugspurger replied:
Happy to be strict for now.

jorisvandenbossche replied:
I made it even more strict (same dtype, not just same class), and added a test for that. 
Will open an issue for the general discussion.

jorisvandenbossche replied:
I opened an issue for this at https://github.com/pandas-dev/pandas/issues/33940

In the end, it seems mainly to come to whether the dtype should exactly be equal or not. 

Since for EAs, the dtype is right now tied to the array class, using equal dtype for now also implies the same class (no additional check to allow sublcasses).

TomAugspurger commented on pandas/tests/extension/base/methods.py:
```diff
@@ -421,3 +421,29 @@ def test_repeat_raises(self, data, repeats, kwargs, error, msg, use_numpy):
             else:
                 data.repeat(repeats, **kwargs)
+
+    def test_equals(self, data, na_value, as_series):
```
Can you also include `DataFrame` in this parameterization, or do it somewhere else? Just to ensure coverage of the changes to `blocks.py`.

TomAugspurger reviewed (commented):
Looks nice. Can you add a release note?

jschendel commented on pandas/core/arrays/base.py:
```diff
@@ -333,6 +335,18 @@ def __iter__(self):
+        """
+        raise AbstractMethodError(self)
+
+    def __ne__(self, other: Any) -> Union[np.ndarray, "ExtensionArray"]:
```
This should allow us to remove the equivalent `IntervalArray` definition here:

https://github.com/pandas-dev/pandas/blob/d149f4109de8610ccc4e3d229c037047b728a49e/pandas/core/arrays/interval.py#L609-L610

jbrockmendel replied:
`Union[np.ndarray, "ExtensionArray"]` -> `ArrayLike`

jorisvandenbossche replied:
@jschendel done!

jbrockmendel commented on pandas/core/arrays/base.py:
```diff
@@ -682,6 +696,33 @@ def searchsorted(self, value, side="left", sorter=None):
+                # boolean array with NA -> fill with False
+                equal_values = equal_values.fillna(False)
+            equal_na = self.isna() & other.isna()
+            return (equal_values | equal_na).all().item()
```
is the `.item()` necessary?

jorisvandenbossche replied:
> is the .item() necessary?

It's to have a python bool, instead of a numpy bool, as result. 
I added a comment to the tests to make it explicit this is the reason we are asserting with `is True/False` (and later on we don't inadvertedly "clean" that up)

jbrockmendel replied:
thanks for clarifying.  i was recently reminded that `np.array(np.timedelta64(1234, "ns")).item()` gives an int instead of timedelta64, so im now cautious around `.item()`

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into gh-27081

jorisvandenbossche committed:
feedback: docs, whatsnew, dataframe test, strict dtype test

jorisvandenbossche committed:
add to reference docs

jorisvandenbossche commented:
@simonjayhawkins how do I solve this typing error?

```
pandas/core/arrays/base.py:338: error: Return type "Union[Any, ExtensionArray]" of "__eq__" incompatible with return type "bool" in supertype "object"
```

So our annotation of `__eq__` is incompatible with the base "object" class, but we do this on purpose. So is there another way to solve this than to suppress the error in some way?

jorisvandenbossche committed:
remove IntervalArray.__ne__

jorisvandenbossche commented:
Any more comments on this PR?

jreback commented:
will look soon

jorisvandenbossche committed:
type ignore following mypy issue (mypy/2783)

jbrockmendel commented on pandas/core/arrays/base.py:
```diff
@@ -335,7 +335,7 @@ def __iter__(self):
             yield self[i]
 
-    def __eq__(self, other: Any) -> ArrayLike:
+    def __eq__(self, other: Any) -> ArrayLike:  # type: ignore[override] # NOQA
```
im not familiar with the `ignore[overrride] # NOQA` pattern.  do we use that elsewhere?

jorisvandenbossche replied:
I thought I posted the mypy issue, but apparently forgot.

So the problem here is that mypy expects a "bool" return for `__eq__` since the base `object` is typed that way in the stubs. In https://github.com/python/mypy/issues/2783, the recommended way to solve this is the above with `# type: ignore[override]`. 

But flake8 doesn't like that, hence also the #NOQA (https://github.com/PyCQA/pyflakes/issues/475). 

This was already done in another PR as well:

https://github.com/pandas-dev/pandas/blob/7aa710a4d4d3b8515a723a366e50f49b327adc23/pandas/core/frame.py#L5062

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/master' into gh-27081

jorisvandenbossche committed:
try again without type: ignore

jorisvandenbossche commented:
@jreback you wanted to take a look here?

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -150,6 +150,7 @@ Other enhancements
   such as ``dict`` and ``list``, mirroring the behavior of :meth:`DataFrame.update` (:issue:`33215`)
 - :meth:`~pandas.core.groupby.GroupBy.transform` and :meth:`~pandas.core.groupby.GroupBy.aggregate` has gained ``engine`` and ``engine_kwargs`` arguments that supports executing functions with ``Numba`` (:issue:`32854`, :issue:`33388`)
 - :meth:`~pandas.core.resample.Resampler.interpolate` now supports SciPy interpolation method :class:`scipy.interpolate.CubicSpline` as method ``cubicspline`` (:issue:`33670`)
+- The ``ExtensionArray`` class has now an ``equals`` method, similarly to ``Series.equals()`` (:issue:`27081`).
```
you can add the doc reference here

jreback commented on pandas/core/internals/blocks.py:
```diff
@@ -1864,6 +1864,11 @@ def where(
         return [self.make_block_same_class(result, placement=self.mgr_locs)]
 
+    def equals(self, other) -> bool:
+        if self.dtype != other.dtype or self.shape != other.shape:
```
why do you need this check here? (its already done on the .values no?)

jorisvandenbossche replied:
Hmm, good question. I just copied that from the base Block.equals method (so it's done there as well). There, `array_equivalent` is used. Maybe that is less strict and the extra check is needed, which is not needed here.

jorisvandenbossche replied:
At least from the docstring of `array_equivalent` (not sure how up to date that is), using that method indeed requires a more strict check in advance of calling it:

https://github.com/pandas-dev/pandas/blob/3ed7dff48bb4e8c7c0129283ff51eccea3a0f861/pandas/core/dtypes/missing.py#L360-L363

But so, for EA.equals that is not needed, so will remove that check here.

jreback reviewed (changes requested):
lgtm. a couple of minor comment.s

jorisvandenbossche committed:
updates

jorisvandenbossche commented:
Travis failures are unrelated (network issues, S3)
